### PR TITLE
support str key type in kvstore

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
Add str->int dict in kvstore backend so that module can use the param name to perform update instead of its index

@mli @piiswrong

### CODE REVIEW

File: python/mxnet/kvstore.py [hunk 1]
```diff
@@ -35,6 +35,32 @@ def _ctype_key_value(keys, vals):
...
+    names = []
+    if isinstance(keys, str):
+        if isinstance(vals, NDArray):
+            names.append(c_str(keys))
+            return (c_array(ctypes.c_char_p, names),
+                    c_array(NDArrayHandle, [vals.handle]))
+        else:
+            for value in vals:
+                assert(isinstance(value, NDArray))
+            return (c_array(ctypes.c_char_p, [c_str(keys)] * len(vals)),
```
Comment (piiswrong) [inline python/mxnet/kvstore.py]: why? I think when vals is a list of NDArray, it means they are duplicates of the same weight on multiple devices

File: python/mxnet/kvstore.py [hunk 2]
```diff
@@ -35,6 +35,32 @@ def _ctype_key_value(keys, vals):
...
+            assert(isinstance(k, str))
+        c_keys = []
+        c_vals = []
+        for key, val in zip(keys, vals):
+            c_key_i, c_val_i = _ctype_str_key_value(key, val)
+            c_keys += c_key_i
+            c_vals += c_val_i
+        return (c_array(ctypes.c_char_p, c_keys), c_array(NDArrayHandle, c_vals))
+
+def _use_str_keys(key):
```
Comment (piiswrong) [inline python/mxnet/kvstore.py]: We can always use str key by converting int to str

Comment (eric-haibin-lin) [inline python/mxnet/kvstore.py]: sure. Do we still want to keep the `MXKVStorePull(..., mx_uint *key)`  API then? Currently I need to create a separate one `MXKVStorePullEx `

Comment (eric-haibin-lin) [inline python/mxnet/kvstore.py]: also the `Pull(vector<int> keys ..)` interface in kvstore.h

Comment (piiswrong) [inline python/mxnet/kvstore.py]: The original API needs to be kept for other packages. Python doesn't need to use it.

Comment (eric-haibin-lin) [inline python/mxnet/kvstore.py]: make sense

File: python/mxnet/kvstore.py [hunk 3]
```diff
@@ -10,31 +10,39 @@
...
-    """
-    Returns ctype arrays for the key-value args. For internal use.
-    """
-    if isinstance(keys, int):
+def _ctype_str_key_value(keys, vals):
+    names = []
+    if isinstance(keys, str):
         if isinstance(vals, NDArray):
-            return (c_array(ctypes.c_int, [keys]),
+            names.append(c_str(keys))
```
Comment (piiswrong) [inline python/mxnet/kvstore.py]: no need for _cast_to_str_keys

always cast to str directly.
names.append(c_str(str(keys)))

Comment (eric-haibin-lin) [inline python/mxnet/kvstore.py]: ok

File: python/mxnet/kvstore.py [hunk 4]
```diff
@@ -11,30 +11,28 @@
 from . import optimizer as opt
 
 def _ctype_key_value(keys, vals):
-    """
-    Returns ctype arrays for the key-value args. For internal use.
-    """
+    names = []
     if isinstance(keys, int):
```
Comment (piiswrong) [inline python/mxnet/kvstore.py]: testing isinstance is not necessary here. str(.) of string is still string. Always convert when appending

Comment (eric-haibin-lin) [inline python/mxnet/kvstore.py]: hmm I changed it yesterday but somehow the commit is lost. Let me push it again

Comment (eric-haibin-lin) [inline python/mxnet/kvstore.py]: well can I just test `isinstance(keys, (tuple, list))` and if it fails, cast single key to `str`?

Comment (piiswrong) [inline python/mxnet/kvstore.py]: yeah, revert the if/else, test for list, tuple first

**Comment by eric-haibin-lin** [comment]
done

**Pull request merged**